### PR TITLE
Merc-1640 fixing Mercury fee calculation

### PR DESCRIPTION
**Pull request by sdrug**

### CODE REVIEW

File: pkg/reportingplugins/mercury/fees.go
Comment (Fletch153) [inline pkg/reportingplugins/mercury/fees.go:22]: A 0 token price is technically correct

File: pkg/reportingplugins/mercury/fees.go
Comment (Fletch153) [inline pkg/reportingplugins/mercury/fees.go:27]: Isn't cents_per_dollar and price_scaling_factor the same thing? it looks like you need to do something like:

baseUSDFeeCents * token.scale / tokenPriceInUsd

File: pkg/reportingplugins/mercury/fees_test.go
Comment (Fletch153) [inline pkg/reportingplugins/mercury/fees_test.go:11]: ditto

File: pkg/reportingplugins/mercury/v2/mercury.go
Comment (Fletch153) [inline pkg/reportingplugins/mercury/v2/mercury.go:196]: Is it possible for the price to be negative? For assets it's unusual, but if there is a bug where the NativePrice.Val is negative, CalculateFee would return negative, and this would be encoded to a uint192, given the most significant is the sign, this would in turn be an extraordinarily large fee and lead to unverifiable reports.

File: pkg/reportingplugins/mercury/fees.go
Comment (sdrug) [reply pkg/reportingplugins/mercury/fees.go:16]: That's a good point. @samsondav how do we want to configure it?

File: pkg/reportingplugins/mercury/fees.go
Comment (sdrug) [reply pkg/reportingplugins/mercury/fees.go:27]: PRICE_SCALING_FACTOR is the multiplier applied to the price we retrieve from reports. 
CENTS_PER_DOLLAR is just to convert fee in cents that we receive from config into usd.

File: pkg/reportingplugins/mercury/v2/mercury.go
Comment (sdrug) [reply pkg/reportingplugins/mercury/v2/mercury.go:196]: Imo handling negative values doesn't belong to the fee calculation and should happen somewhere upstream. Also having unverifiable report in this case sounds like a good outcome to me. I'll let @samsondav to chime in here as the code owner

File: pkg/reportingplugins/mercury/fees.go
Comment (Fletch153) [reply pkg/reportingplugins/mercury/fees.go:27]: I believe you only need one, and that is the precision of the token

File: pkg/reportingplugins/mercury/v2/mercury.go
Comment (Fletch153) [reply pkg/reportingplugins/mercury/v2/mercury.go:196]: If we have a bug in the software which leads to int_max fees being generated, there will be a complete outage. I think instead, we should remove the concept of int_max for failures and instead, zero the fee.

This also reduces gas costs quite significantly, as passing 0x00..00 is cheaper than passing 0x_int_max not only because 0 is billed for less, but low entropy data can easily be compressed when rolling up on a L2

File: pkg/reportingplugins/mercury/fees.go
Comment (samsondav) [reply pkg/reportingplugins/mercury/fees.go:16]: For now we can leave it hardcoded since both link and eth use 18 dp

File: pkg/reportingplugins/mercury/fees.go
Comment (samsondav) [inline pkg/reportingplugins/mercury/fees.go:25]: For prices please do not use float, use decimal for absolute precision

File: pkg/reportingplugins/mercury/v2/mercury.go
Comment (samsondav) [reply pkg/reportingplugins/mercury/v2/mercury.go:196]: Core side will never encode a negative fee. But it doesn't hurt to have extra checks.

**Review by samsondav — CHANGES_REQUESTED** [review]

File: pkg/reportingplugins/mercury/fees_test.go
Comment (samsondav) [inline pkg/reportingplugins/mercury/fees_test.go:47]: Let's sanity check handling negative base fee + negative token price as well pls, should return 0 in these cases or error

**Review by samsondav — CHANGES_REQUESTED** [review]

**Review by samsondav — APPROVED** [review]

**Pull request merged**
